File: apps/android/app/src/main/java/site/smap/harubook/features/reader/ReaderViewModel.kt
package site.smap.harubook.features.reader

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import site.smap.harubook.core.audio.AudioPlayer
import site.smap.harubook.core.models.Book
import site.smap.harubook.core.models.BookDetail
import site.smap.harubook.core.models.Mission
import site.smap.harubook.core.models.Passage
import site.smap.harubook.core.models.ReadingLogResponse
import site.smap.harubook.core.models.VocabularyEntry
import site.smap.harubook.core.networking.ApiClient
import site.smap.harubook.core.srs.SrsGrade
import site.smap.harubook.core.srs.SrsStore
import site.smap.harubook.core.srs.srsNormalizeKey

/**
 * 본문 텍스트 크기. SharedPreferences로 사용자 전역 선호도 영속화.
 */
enum class ReaderTextScale(val sp: Int, val label: String, val previewSp: Int) {
    Small(22, "작게", 12),
    Medium(28, "보통", 15),
    Large(34, "크게", 18),
    XLarge(40, "아주 크게", 22);

    companion object {
        private const val FILE = "harubook_reader"
        private const val KEY = "textScale"

        fun load(context: Context): ReaderTextScale {
            val raw = context.applicationContext
                .getSharedPreferences(FILE, Context.MODE_PRIVATE)
                .getString(KEY, null) ?: return Medium
            return runCatching { valueOf(raw) }.getOrDefault(Medium)
        }

        fun save(context: Context, value: ReaderTextScale) {
            context.applicationContext
                .getSharedPreferences(FILE, Context.MODE_PRIVATE)
                .edit().putString(KEY, value.name).apply()
        }
    }
}

data class ReaderUiState(
    /** 책 메타데이터 — 헤더에 제목/나이 배지/CEFR 배지를 표시하려면 필수. iOS 패리티(viewModel.book). */
    val book: Book? = null,
    val passages: List<Passage> = emptyList(),
    val vocabulary: List<VocabularyEntry> = emptyList(),
    val currentIndex: Int = 0,
    val showsKorean: Boolean = false,
    val isLoadingDetail: Boolean = true,
    val error: String? = null,
    val textScale: ReaderTextScale = ReaderTextScale.Medium,
    val readingLogId: Int? = null,
    val synthesizingPassageId: Int? = null,
    val generatingScenePassageId: Int? = null,
    /** 사용자가 본문에서 탭한 vocab. null이면 popover 닫힘. */
    val selectedVocab: VocabularyEntry? = null,
    /** passageIndex → 검증된 미션. 레거시 책(missions 부재)은 빈 맵 — 미션 UI 없음. */
    val missionByIndex: Map<Int, Mission> = emptyMap(),
    /** 완료한 미션의 passageIndex 집합. SharedPreferences로 복원/영속화. */
    val completedMissions: Set<Int> = emptySet(),
)

/**
 * iOS `ReaderViewModel.swift` 미러.
 *
 * 두 가지 백엔드 흐름을 함께 다룬다:
 *  - `/api/books/{id}` 로 본문/오디오/장면 메타데이터 로드
 *  - `/api/logs` POST/PATCH 로 reading_logs 진행률·완료 시각 누적 갱신
 *
 * 그리고 passage 오디오 부재 시 `/api/tts/{id}` 합성 트리거, 장면 이미지 부재 시
 * `/api/image/passage/{id}` 트리거.
 */
class ReaderViewModel(
    private val bookId: Int,
    private val profileId: Int,
    private val appContext: Context,
) : ViewModel() {

    private val _state = MutableStateFlow(
        ReaderUiState(
            textScale = ReaderTextScale.load(appContext),
            completedMissions = loadCompletedMissions(),
        ),
    )
    val state: StateFlow<ReaderUiState> = _state.asStateFlow()

    private val srs: SrsStore = SrsStore.create(appContext, profileId)
    private var hasReportedFinish = false

    fun bootstrap() {
        viewModelScope.launch {
            coroutineScope {
                awaitAll(async { loadDetail() }, async { startLog() })
            }
        }
    }

    fun reportPageChanged(newIndex: Int) {
        viewModelScope.launch {
            val current = _state.value
            // 페이지 전환 시 한글 해석 자동 닫기 — 다음 문장은 영문부터.
            if (newIndex != current.currentIndex) {
                _state.update { it.copy(showsKorean = false) }
            }
            _state.update { it.copy(currentIndex = newIndex) }
            val passages = _state.value.passages
            // 떠나는 페이지의 오디오 정지 — 문장이 3~6문장으로 길어져(오디오 30초+) 이전
            // 낭독이 다음 페이지까지 이어지면 "정지가 안 된다"는 혼란을 만든다. iOS pageChanged 패리티.
            val playingId = AudioPlayer.state.value.nowPlayingPassageId
            if (playingId != null && passages.getOrNull(newIndex)?.id != playingId) {
                AudioPlayer.stop()
            }
            if (passages.isEmpty()) return@launch
            val total = passages.size.coerceAtLeast(1)
            val ratio = (newIndex + 1).toDouble() / total.toDouble()
            patchLog(progressRatio = ratio, finishedAtUnix = null)
        }
    }

    fun leave() {
        if (hasReportedFinish) return
        hasReportedFinish = true
        val now = System.currentTimeMillis() / 1000
        viewModelScope.launch {
            patchLog(progressRatio = null, finishedAtUnix = now)
            AudioPlayer.stop()
        }
    }

    fun toggleKorean() {
        _state.update { it.copy(showsKorean = !it.showsKorean) }
    }

    fun setTextScale(scale: ReaderTextScale) {
        ReaderTextScale.save(appContext, scale)
        _state.update { it.copy(textScale = scale) }
    }

    /** 재생 토글. audioPath 부재 시 `/api/tts/{id}` 합성 먼저. */
    fun togglePlayback(passageIndex: Int) {
        val passage = _state.value.passages.getOrNull(passageIndex) ?: return
        val existing = passage.audioPath
        if (!existing.isNullOrEmpty()) {
            AudioPlayer.toggle(passage.id, existing)
            return
        }
        viewModelScope.launch {
            _state.update { it.copy(synthesizingPassageId = passage.id) }
            try {
                // passage가 3~6문장(최대 ~95단어)으로 길어져 합성이 기본 60s에
                // 근접할 수 있어 90s로 상향. iOS ReaderViewModel과 패리티.
                val response: TtsResponse = ApiClient.post(
                    path = "/api/tts/${passage.id}",
                    timeoutMillis = 90_000,
                )
                _state.update { st ->
                    st.copy(
                        passages = st.passages.map { p ->
                            if (p.id == passage.id) p.copy(audioPath = response.audioPath) else p
                        },
                        synthesizingPassageId = null,
                    )
                }
                AudioPlayer.toggle(passage.id, response.audioPath)
            } catch (e: Throwable) {
                _state.update {
                    it.copy(
                        synthesizingPassageId = null,
                        error = "오디오 준비 실패: ${e.message ?: e::class.java.simpleName}",
                    )
                }
            }
        }
    }

    /** 장면 이미지 합성(멱등). */
    fun requestSceneImage(passageIndex: Int) {
        val passage = _state.value.passages.getOrNull(passageIndex) ?: return
        if (!passage.sceneImagePath.isNullOrEmpty()) return

        viewModelScope.launch {
            _state.update { it.copy(generatingScenePassageId = passage.id) }
            try {
                val response: SceneImageResponse = ApiClient.post(path = "/api/image/passage/${passage.id}")
                _state.update { st ->
                    st.copy(
                        passages = st.passages.map { p ->
                            if (p.id == passage.id) p.copy(sceneImagePath = response.sceneImagePath) else p
                        },
                        generatingScenePassageId = null,
                    )
                }
            } catch (e: Throwable) {
                _state.update {
                    it.copy(
                        generatingScenePassageId = null,
                        error = "삽화 생성 실패: ${e.message ?: e::class.java.simpleName}",
                    )
                }
            }
        }
    }

    fun selectVocab(entry: VocabularyEntry) {
        _state.update { it.copy(selectedVocab = entry) }
    }

    fun dismissVocab() {
        _state.update { it.copy(selectedVocab = null) }
    }

    fun gradeVocab(entry: VocabularyEntry, grade: SrsGrade) {
        srs.grade(entry.word, grade)
        _state.update { it.copy(selectedVocab = null) }
    }

    // MARK: - 책 속 미션 (웹 reader.tsx missionByIdx/handleWordTap/completeMission 패리티)

    /**
     * 워드 헌트 판정 — 해당 passage에 미완료 워드 헌트가 있고, 탭한 단어가 targetWord와
     * 일치하면 완료. 그 외의 단어 탭은 기존 뜻 보기 동작 그대로(이 함수는 부수 판정만).
     * normalize는 SRS와 동일한 trim+lowercase+구두점 제거([srsNormalizeKey]) 재사용.
     */
    fun reportWordTapped(passageIndex: Int, word: String) {
        val current = _state.value
        val hunt = current.missionByIndex[passageIndex]?.wordHunt ?: return
        if (current.completedMissions.contains(passageIndex)) return
        if (srsNormalizeKey(word) == srsNormalizeKey(hunt.targetWord)) {
            completeMission(passageIndex)
        }
    }

    /** 미션 완료 — 상태 갱신 + SharedPreferences 영속화(웹 localStorage missionKey 패리티). */
    fun completeMission(passageIndex: Int) {
        if (_state.value.completedMissions.contains(passageIndex)) return
        val next = _state.value.completedMissions + passageIndex
        _state.update { it.copy(completedMissions = next) }
        persistCompletedMissions(next)
    }

    /** ReaderTextScale과 같은 리더 전용 prefs 파일. */
    private fun readerPrefs() =
        appContext.applicationContext.getSharedPreferences("harubook_reader", Context.MODE_PRIVATE)

    /** 웹 localStorage 키(`reader:mission:{bookId}`)와 동일한 네이밍 — 정수 배열 JSON. */
    private fun missionStorageKey(): String = "reader:mission:$bookId"

    /** 완료 미션 복원 — 정수 배열(JSON)만 신뢰, 손상 시 빈 집합(fail-soft). */
    private fun loadCompletedMissions(): Set<Int> {
        val raw = readerPrefs().getString(missionStorageKey(), null) ?: return emptySet()
        return runCatching {
            Json.decodeFromString(ListSerializer(Int.serializer()), raw).toSet()
        }.getOrDefault(emptySet())
    }

    private fun persistCompletedMissions(done: Set<Int>) {
        readerPrefs().edit()
            .putString(
                missionStorageKey(),
                Json.encodeToString(ListSerializer(Int.serializer()), done.toList()),
            )
            .apply()
    }

    /**
     * passageIndex → Mission 맵 — 웹 reader.tsx missionByIdx 패리티.
     * 저장 시 서버가 범위/단어 존재를 이미 검증했지만(fail-soft), 레거시/수동 편집 대비
     *  - passageIndex 범위 밖 → 무시
     *  - 워드 헌트는 vocabulary에 있는 단어일 때만 유효(탭 대상이 밑줄 단어뿐이므로)
     *  - check는 질문/2지선다가 온전할 때만 유효(부분 파싱된 미션이 빈 카드로 뜨지 않게)
     */
    private fun buildMissionMap(
        missions: List<Mission>?,
        vocabulary: List<VocabularyEntry>,
        passageCount: Int,
    ): Map<Int, Mission> {
        if (missions.isNullOrEmpty()) return emptyMap()
        val vocabMap = buildVocabMap(vocabulary)
        val map = LinkedHashMap<Int, Mission>()
        for (m in missions) {
            if (m.passageIndex < 0 || m.passageIndex >= passageCount) continue
            val hunt = m.wordHunt?.takeIf {
                it.targetWord.isNotBlank() && vocabMap.containsKey(normalizeVocabKey(it.targetWord))
            }
            val check = m.check?.takeIf {
                it.question.isNotBlank() && it.choices.size >= 2 && it.answerIndex in it.choices.indices
            }
            if (hunt == null && check == null) continue
            map[m.passageIndex] = m.copy(wordHunt = hunt, check = check)
        }
        return map
    }

    private suspend fun loadDetail() {
        try {
            val detail: BookDetail = ApiClient.get(path = "/api/books/$bookId")
            val sorted = detail.passages.sortedBy { p -> p.orderIndex }
            _state.update {
                it.copy(
                    book = detail.book,
                    passages = sorted,
                    vocabulary = detail.book.vocabulary.orEmpty(),
                    missionByIndex = buildMissionMap(
                        missions = detail.book.missions,
                        vocabulary = detail.book.vocabulary.orEmpty(),
                        passageCount = sorted.size,
                    ),
                    isLoadingDetail = false,
                )
            }
        } catch (e: Throwable) {
            _state.update {
                it.copy(
                    isLoadingDetail = false,
                    error = e.message ?: "본문을 불러오지 못했어요.",
                )
            }
        }
    }

    private suspend fun startLog() {
        try {
            val response: ReadingLogResponse = ApiClient.post(
                path = "/api/logs",
                body = StartLogRequest(profileId = profileId, bookId = bookId),
            )
            _state.update { it.copy(readingLogId = response.log.id) }
        } catch (_: Throwable) {
            // 소프트 페일.
        }
    }

    private suspend fun patchLog(progressRatio: Double?, finishedAtUnix: Long?) {
        val logId = _state.value.readingLogId ?: return
        try {
            ApiClient.patch<ReadingLogResponse>(
                path = "/api/logs",
                body = PatchLogRequest(
                    id = logId,
                    progressRatio = progressRatio,
                    finishedAtUnix = finishedAtUnix,
                    quizScore = null,
                ),
            )
        } catch (_: Throwable) {
            // 소프트 페일.
        }
    }
}

@Serializable
internal data class StartLogRequest(val profileId: Int, val bookId: Int)

@Serializable
internal data class PatchLogRequest(
    val id: Int,
    val progressRatio: Double? = null,
    val finishedAtUnix: Long? = null,
    val quizScore: Int? = null,
)

@Serializable
internal data class TtsResponse(
    val passageId: Int,
    val audioPath: String,
    val cached: Boolean? = null,
    val bytes: Int? = null,
)

@Serializable
internal data class SceneImageResponse(
    val passageId: Int? = null,
    val sceneImagePath: String,
)

File: apps/ios/HaruBook/Features/Reader/ReaderViewModel.swift
import Foundation
import Observation
import CoreGraphics

/// 본문 텍스트 크기. 사용자 전역 선호도로 UserDefaults에 영속화.
enum ReaderTextScale: String, CaseIterable, Identifiable {
    case small, medium, large, xlarge

    var id: String { rawValue }

    /// 본문 폰트 사이즈(`atozRegular`). medium = 28pt를 기본값으로 — 어린이 학습 상황에서
    /// 22pt는 너무 작다는 피드백을 반영. 더 작게/더 크게 양방향 옵션 유지.
    var fontSize: CGFloat {
        switch self {
        case .small:  return 22
        case .medium: return 28
        case .large:  return 34
        case .xlarge: return 40
        }
    }

    var label: String {
        switch self {
        case .small:  return "작게"
        case .medium: return "보통"
        case .large:  return "크게"
        case .xlarge: return "아주 크게"
        }
    }

    /// 텍스트 크기 컨트롤의 "A" 미리보기 글자 크기 — 단계 차이를 시각적으로 보여주기 위한 값.
    /// 실제 본문 폰트(`fontSize`)와는 별개.
    var controlPreviewSize: CGFloat {
        switch self {
        case .small:  return 12
        case .medium: return 15
        case .large:  return 18
        case .xlarge: return 22
        }
    }

    private static let defaultsKey = "readerTextScale"

    static func load() -> ReaderTextScale {
        let raw = UserDefaults.standard.string(forKey: defaultsKey) ?? ""
        return ReaderTextScale(rawValue: raw) ?? .medium
    }

    func save() {
        UserDefaults.standard.set(rawValue, forKey: Self.defaultsKey)
    }
}

@Observable
@MainActor
final class ReaderViewModel {
    let book: Book
    let profileId: Int

    var passages: [Passage] = []
    var currentIndex: Int = 0
    var showsKorean: Bool = false
    var isLoadingDetail: Bool = true
    var error: String?
    /// 본문 텍스트 크기. 변경 시 UserDefaults에 영속화 — 모든 책에서 같은 크기로 시작.
    var textScale: ReaderTextScale = ReaderTextScale.load() {
        didSet { textScale.save() }
    }

    /// 현재 진행 중인 reading_log id. 외부(QuizView)가 점수 PATCH 시 참조한다.
    private(set) var readingLogId: Int?

    /// 책 속 미션 — passageIndex → Mission. 서버가 저장 시 검증하지만(fail-soft) 레거시/수동
    /// 편집 대비 워드 헌트는 vocabulary에 있는 단어일 때만 유효로 간주한다(웹 reader의
    /// missionByIdx와 동일). 레거시 책(missions nil)은 빈 맵 — 미션 UI 없이 렌더.
    let missionsByIndex: [Int: Mission]

    /// 완료한 미션의 passageIndex 집합 — UserDefaults 복원(웹 localStorage `reader:mission:{bookId}` 패리티).
    private(set) var missionsDone: Set<Int>

    /// 마지막으로 TTS 합성이 진행 중인 passage id. UI 인디케이터용.
    private(set) var synthesizingPassageId: Int?

    /// 장면 이미지 생성이 진행 중인 passage id. UI 인디케이터용.
    private(set) var generatingScenePassageId: Int?

    private var hasReportedFinish: Bool = false

    init(book: Book, profileId: Int) {
        self.book = book
        self.profileId = profileId
        self.missionsByIndex = Self.buildMissionIndex(book: book)
        self.missionsDone = Self.loadMissionsDone(bookId: book.id)
    }

    func bootstrap() async {
        async let detail: Void = loadDetail()
        async let log: Void = startLog()
        _ = await (detail, log)
    }

    /// 페이지 전환 — 반드시 동기로 상태를 갱신한다.
    ///
    /// TabView(selection:)의 setter가 비동기(Task)로만 상태를 바꾸면 SwiftUI가 바인딩
    /// 값 불일치를 감지해 페이지가 되돌아가거나(스냅백) 스와이프가 씹히는 오작동이 난다.
    /// 진행률 서버 PATCH만 비동기로 분리.
    func pageChanged(to newIndex: Int) {
        // 페이지 전환 시 한글 해석은 자동으로 닫는다 — 다음 문장은 영문부터 다시 만나도록 학습 흐름 유지.
        // 사용자가 다시 필요하면 한글 토글 버튼으로 켤 수 있다.
        if newIndex != currentIndex {
            showsKorean = false
        }
        currentIndex = newIndex
        // 떠나는 페이지의 오디오 정지 — 문장이 3~6문장으로 길어져(오디오 30초+) 이전
        // 낭독이 다음 페이지까지 이어지면 "정지가 안 된다"는 혼란을 만든다.
        if let playingId = AudioPlayer.shared.nowPlayingPassageId,
           passages.indices.contains(newIndex),
           passages[newIndex].id != playingId {
            AudioPlayer.shared.stop()
        }
        guard !passages.isEmpty else { return }
        let total = max(passages.count, 1)
        let ratio = Double(newIndex + 1) / Double(total)
        Task { await patchLog(progressRatio: ratio, finishedAtUnix: nil) }
    }

    func leave() async {
        guard !hasReportedFinish else { return }
        hasReportedFinish = true
        let now = Int(Date().timeIntervalSince1970)
        await patchLog(progressRatio: nil, finishedAtUnix: now)
        AudioPlayer.shared.stop()
    }

    func toggleKorean() {
        showsKorean.toggle()
    }

    // MARK: - 책 속 미션

    func mission(for passageIndex: Int) -> Mission? {
        missionsByIndex[passageIndex]
    }

    /// 워드 헌트 판정 — 해당 passage에 미완료 워드 헌트가 있고 탭한 단어가 targetWord와
    /// 일치하면 완료. 그 외의 단어 탭은 기존 뜻 보기 동작 그대로(여기서는 아무것도 안 함).
    /// 정규화(trim+lowercase+구두점 제거)는 SRS와 같은 규칙 — srsNormalizeKey 재사용.
    func handleWordTap(_ word: String, passageIndex: Int) {
        guard let hunt = missionsByIndex[passageIndex]?.wordHunt,
              !missionsDone.contains(passageIndex) else { return }
        if srsNormalizeKey(word) == srsNormalizeKey(hunt.targetWord) {
            completeMission(at: passageIndex)
        }
    }

    /// 미션 완료 — 로컬 영속화(정수 배열). 진행을 막지 않는 재미 요소라 서버 기록 없음(웹과 동일).
    func completeMission(at passageIndex: Int) {
        guard !missionsDone.contains(passageIndex) else { return }
        missionsDone.insert(passageIndex)
        UserDefaults.standard.set(
            Array(missionsDone).sorted(),
            forKey: Self.missionDefaultsKey(bookId: book.id),
        )
    }

    private static func missionDefaultsKey(bookId: Int) -> String {
        "reader:mission:\(bookId)"
    }

    private static func buildMissionIndex(book: Book) -> [Int: Mission] {
        guard let missions = book.missions, !missions.isEmpty else { return [:] }
        let vocabKeys = Set((book.vocabulary ?? []).map { srsNormalizeKey($0.word) })
        var map: [Int: Mission] = [:]
        for m in missions {
            // lossy 디코딩 강등(-1) 포함, 음수 인덱스는 폐기. 상한 초과는 조회에 안 걸려 자연 무시.
            guard m.passageIndex >= 0 else { continue }
            // 워드 헌트는 탭 대상이 밑줄(vocabulary) 단어뿐이므로 vocabulary에 없으면 무시.
            var hunt: MissionWordHunt?
            if let wordHunt = m.wordHunt, vocabKeys.contains(srsNormalizeKey(wordHunt.targetWord)) {
                hunt = wordHunt
            }
            // check는 질문/선택지/정답 인덱스가 온전할 때만 유효 — 불완전 미션이
            // 버튼 없는 빈 카드로 뜨지 않게(AOS buildMissionMap과 대칭).
            var check: MissionCheck?
            if let candidate = m.check,
               !candidate.question.isEmpty,
               candidate.choices.count >= 2,
               candidate.choices.indices.contains(candidate.answerIndex) {
                check = candidate
            }
            // wordHunt/check 둘 다 비면 미션 자체를 버린다.
            guard hunt != nil || check != nil else { continue }
            map[m.passageIndex] = Mission(passageIndex: m.passageIndex, wordHunt: hunt, check: check)
        }
        return map
    }

    private static func loadMissionsDone(bookId: Int) -> Set<Int> {
        let raw = UserDefaults.standard.array(forKey: missionDefaultsKey(bookId: bookId)) ?? []
        return Set(raw.compactMap { $0 as? Int })
    }

    /// 재생 토글. audioPath가 없으면 TTS 합성을 먼저 요청한다.
    func togglePlayback(for passageIndex: Int) async {
        guard passages.indices.contains(passageIndex) else { return }
        let passage = passages[passageIndex]

        if let path = passage.audioPath, !path.isEmpty {
            AudioPlayer.shared.toggle(passageId: passage.id, audioPath: path)
            return
        }

        synthesizingPassageId = passage.id
        defer { synthesizingPassageId = nil }
        do {
            // passage가 3~6문장(최대 ~95단어)으로 길어져 합성이 기본 60s에
            // 근접할 수 있어 90s로 상향. Android ReaderViewModel과 패리티.
            let response: TtsResponse = try await APIClient.shared.send(
                Endpoint(path: "/api/tts/\(passage.id)", method: .post, requiresAuth: true, timeout: 90)
            )
            if let idx = passages.firstIndex(where: { $0.id == passage.id }) {
                passages[idx] = Passage(
                    id: passage.id,
                    bookId: passage.bookId,
                    orderIndex: passage.orderIndex,
                    textEn: passage.textEn,
                    textKo: passage.textKo,
                    audioPath: response.audioPath,
                    sceneImagePath: passage.sceneImagePath
                )
            }
            AudioPlayer.shared.toggle(passageId: passage.id, audioPath: response.audioPath)
        } catch {
            self.error = error.localizedDescription
        }
    }

    /// 장면 이미지를 합성한다(`POST /api/image/passage/[id]`). 멱등.
    func requestSceneImage(for passageIndex: Int) async {
        guard passages.indices.contains(passageIndex) else { return }
        let passage = passages[passageIndex]
        if passage.sceneImagePath?.isEmpty == false { return }

        generatingScenePassageId = passage.id
        defer { generatingScenePassageId = nil }
        do {
            let response: SceneImageResponse = try await APIClient.shared.send(
                Endpoint(path: "/api/image/passage/\(passage.id)", method: .post)
            )
            if let idx = passages.firstIndex(where: { $0.id == passage.id }) {
                passages[idx] = Passage(
                    id: passage.id,
                    bookId: passage.bookId,
                    orderIndex: passage.orderIndex,
                    textEn: passage.textEn,
                    textKo: passage.textKo,
                    audioPath: passage.audioPath,
                    sceneImagePath: response.sceneImagePath
                )
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    // MARK: - Private

    private func loadDetail() async {
        do {
            let detail: BookDetail = try await APIClient.shared.send(
                Endpoint(path: "/api/books/\(book.id)")
            )
            self.passages = detail.passages.sorted(by: { $0.orderIndex < $1.orderIndex })
            self.isLoadingDetail = false
        } catch {
            self.error = error.localizedDescription
            self.isLoadingDetail = false
        }
    }

    private func startLog() async {
        do {
            let response: ReadingLogResponse = try await APIClient.shared.send(
                Endpoint(
                    path: "/api/logs",
                    method: .post,
                    body: StartLogRequest(profileId: profileId, bookId: book.id)
                )
            )
            self.readingLogId = response.log.id
        } catch {
            // 소프트 페일
        }
    }

    private func patchLog(progressRatio: Double?, finishedAtUnix: Int?) async {
        guard let logId = readingLogId else { return }
        do {
            let _: ReadingLogResponse = try await APIClient.shared.send(
                Endpoint(
                    path: "/api/logs",
                    method: .patch,
                    body: PatchLogRequest(
                        id: logId,
                        progressRatio: progressRatio,
                        finishedAtUnix: finishedAtUnix,
                        quizScore: nil
                    )
                )
            )
        } catch {
            // 소프트 페일
        }
    }
}

private struct StartLogRequest: Encodable {
    let profileId: Int
    let bookId: Int
}

private struct PatchLogRequest: Encodable {
    let id: Int
    let progressRatio: Double?
    let finishedAtUnix: Int?
    let quizScore: Int?
}

private struct TtsResponse: Decodable {
    let passageId: Int
    let audioPath: String
    let cached: Bool?
    let bytes: Int?
}

private struct SceneImageResponse: Decodable {
    let passageId: Int?
    let sceneImagePath: String
}
